### ISSUE TITLE
Sometimes dgn deletes too much

### ISSUE THREAD

**Issue by Raimondi** [vim/vim#934]
Steps:
- Start vim
- Insert the following text: 123
- Search with /\d/
- Set cursor on the "2"
- Type dgn

Expected:

```
13
```

Result:

```
1
```

There can be any number of characters between the numbers and  3 can be in a different line, but 1 and 2 must be on the same line.
`git bisect` says the issue started with @chrisbra's fix for #295:

```
b12db9f92ea74039bb0b4110add4232f0bf338ea is the first bad commit
commit b12db9f92ea74039bb0b4110add4232f0bf338ea
Author: Bram Moolenaar <Bram@vim.org>
Date:   Sat Dec 13 22:00:22 2014 +0100

    updated for version 7.4.551
    Problem:    "ygn" may yank too much. (Fritzophrenic)  Issue 295.
    Solution:   Check the width of the next match. (Christian Brabandt)
```

**Comment by chrisbra** [comment]
I think, this patch [here](https://groups.google.com/d/msg/vim_dev/kT0XucEJsng/WcFcddu_BAAJ) fixes it. Can you confirm? Bram, would you consider merging if it is confirmed to be fixed, as I have posted before and it includes a test?

**Comment by Raimondi** [comment]
The patch works for me. But it fails to apply all changes in testdir:

```
$ cat gn_bug.diff | patch -p1
(Stripping trailing CRs from patch; use --binary to disable.)
patching file src/search.c
Hunk #1 succeeded at 4719 (offset -39 lines).
Hunk #2 succeeded at 4734 (offset -41 lines).
Hunk #3 succeeded at 4753 (offset -41 lines).
(Stripping trailing CRs from patch; use --binary to disable.)
patching file src/testdir/Make_all.mak
Hunk #1 FAILED at 173.
1 out of 1 hunk FAILED -- saving rejects to file src/testdir/Make_all.mak.rej
(Stripping trailing CRs from patch; use --binary to disable.)
patching file src/testdir/test53.in
(Stripping trailing CRs from patch; use --binary to disable.)
patching file src/testdir/test53.ok
(Stripping trailing CRs from patch; use --binary to disable.)
patching file src/testdir/test_gn.vim
```

```
$ cat src/testdir/Make_all.mak.rej
--- src/testdir/Make_all.mak
+++ src/testdir/Make_all.mak
@@ -173,6 +173,7 @@
 # Keep test_alot.res as the last one, sort the others.
 NEW_TESTS = test_assert.res \
        test_cdo.res \
+       test_gn.res \
        test_increment.res \
        test_quickfix.res \
        test_viml.res \
```

**Comment by brammool** [comment]
Christian Brabandt wrote:

> I think, this patch
> [here](https://groups.google.com/d/msg/vim_dev/kT0XucEJsng/WcFcddu_BAAJ)
> fixes it. Can you confirm? Bram, would you consider merging if it is
> confirmed to be fixed, as I have posted before and it includes a test?

Is it this entry in the todo list:

Half-finished patch to fix the Problem using cgn to change a search hit when
replacement includes hit. Reported by John Beckett, fix by Christian Brabandt,
2016 Jan 11.

Not sure why it says "Half-finished".

## 

A poem:                read aloud:

<> !_''#               Waka waka bang splat tick tick hash,
^"`$$-                 Caret quote back-tick dollar dollar dash,
!_=@$_                 Bang splat equal at dollar under-score,
%*<> ~#4               Percent splat waka waka tilde number four,
&[]../                 Ampersand bracket bracket dot dot slash,
|{,,SYSTEM HALTED      Vertical-bar curly-bracket comma comma CRASH.

Fred Bremmer and Steve Kroese (Calvin College & Seminary of Grand Rapids, MI.)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
On Di, 19 Jul 2016, Bram Moolenaar wrote:

> Christian Brabandt wrote:
> 
> > I think, this patch
> > [here](https://groups.google.com/d/msg/vim_dev/kT0XucEJsng/WcFcddu_BAAJ)
> > fixes it. Can you confirm? Bram, would you consider merging if it is
> > confirmed to be fixed, as I have posted before and it includes a test?
> 
> Is it this entry in the todo list:
> 
> Half-finished patch to fix the Problem using cgn to change a search hit when
> replacement includes hit. Reported by John Beckett, fix by Christian Brabandt,
> 2016 Jan 11.

Yes, that would be it.

> Not sure why it says "Half-finished".

I don't know why.

Mit freundlichen Grüßen

## Christian

Man sollte niemand über Furchtsamkeit tadeln, bis man weiß, wie wenig
oder viel er dagegen gearbeitet.
        -- Jean Paul

**Comment by chrisbra** [comment]
Fixed by [7.4.2100](https://github.com/vim/vim/releases/tag/v7.4.2100)
